Zephyr 3.5 update

petejohanson opened the pull request:
Even more experimental than #1853 as this one has even less testing, but opening this for posterity. Still needs me to publish newer Docker images for CI to be happy, as this needs a newer SDK version.

Blog Post preview: https://deploy-preview-1995--zmk.netlify.app/blog/2024/01/31/zephyr-3-5

# Testing steps

## Local dev

1. If not building w/ VSCode docker integration, install https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.16.4
1. `gh pr checkout 1995`
2. `west update`
3. Use new build directory to build ZMK as usual.

## GitHub Actions

Make two edits in the branch of your config repo:

1. ~~Update `.github/workflowr/build.yaml` to contain:~~ No longer needed, the `stable` Docker image has been bumped so this step can be skipped now.

    ```
    on: [push, pull_request, workflow_dispatch]

    jobs:
      build:
        uses: petejohanson/zmk/.github/workflows/build-user-config.yml@core/zephyr-3.5-update
    ```
 1. Update `config/west.yaml` according to the https://zmk.dev/docs/features/beta-testing/ instructions to test this PR.

## To Do

- [x] https://github.com/zmkfirmware/zmk/pull/1853#issuecomment-1789915446
- [x] Blog post with upgrade details
- [x] Get a the retained bootmode change working.
- [x] Tag the proper SDK zmk-docker images and move to using them.
- [x] Lots of testing

petejohanson commented:
I addressed the RGB funkiness with SPI pin setup on nRF52 with this change which I added to our Zephyr fork and submitted upstream: https://github.com/zephyrproject-rtos/zephyr/pull/67341 ( /cc @xudongzheng )

petejohanson commented:
> Will `behavior_mouse_key_press.c` also need to be switched to `POST_KERNEL`?
> 
> https://github.com/zmkfirmware/zmk/blob/40c9d9fc5993f189e55403a355aec96a5c8d8aae/app/src/behaviors/behavior_mouse_key_press.c#L42

Indeed. Missed this, thanks.

> 
> The DTS binding for the `MAX17048` sensor (...fuel gauge?) still [collides with Zephyr's](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516169561/job/20460957357#step:11:33):
> 
> ```
> devicetree error: both /home/less/osrc/zmk-src/zmk/app/module/dts/bindings/sensor/maxim,max17048.yml and /home/less/osrc/zmk-src/zmk/zephyr/dts/bindings/fuel-gauge/maxim,max17048.yaml have 'compatible: maxim,max17048' and 'on-bus: i2c'
> ```
> 
> If I [delete ZMK's DTS binding](https://github.com/lesshonor/zmk/commit/3d0d2a6de5568e918e3ad324981c92d8ab7c717f), I get a number of different results depending upon where and in what way this thing is enabled; no idea which of these outcomes is correct, if any.
> 
> 1. [Using `MAX17048` in Kconfig.defconfig](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516258743/job/20461198917) results in a successful compilation that _appears_ to use ZMK's driver exclusively.
> 2. [Using `ZMK_MAX17048` in Kconfig.defconfig](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516314666/job/20461242609) (or `CONFIG_ZMK_MAX17048=y` in `uwu30_defconfig`) fails at the linker; it seemingly can't find the `zmk,battery`-compatible node?
> 3. [Using `CONFIG_FUEL_GAUGE=y` and `CONFIG_MAX17048=y` in `uwu30_defconfig`](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516310376/job/20461236910) instead also fails at the linker, but for a different reason: it pulls in both the ZMK and Zephyr drivers??
> 
> ...I'm sure some of this is my not understanding Zephyr. I have no way to test any changes on hardware yet either, so I can't determine if scenario 1 works as intended by the board's creator.
> 
> (paging @zhiayang who may be able to test on actual hardware, given #1946)

You need to `west update` after checking out the branch. I'll be prepping testing notes soon that include all these details.

zhiayang commented:
huh. what will we do with the existing max17048 driver now that upstream supports it? just yeet it?

petejohanson commented:
> huh. what will we do with the existing max17048 driver now that upstream supports it? just yeet it?

We can get rid of ours, but we need code that consumes the new fuel gauge API instead of a sensor before we can do that.

zhiayang commented:
mm, makes sense 👍🏻

bryanforbes commented on docs/blog/2024-01-31-zephyr-3-5.md:
```diff
@@ -0,0 +1,134 @@
+
+### User Config Repositories Using GitHub Actions
+
+Existing user config repositories using Github Actions to build will pull down Zephyr 3.2 automatically, however if you created your user config a while ago, you may need to update it to reference our shared build configuration to leverage the correct Docker image.
```
I believe this should say "Zephyr 3.5"

petejohanson replied:
Indeed. Fixed.

caksoylar commented on docs/blog/2024-01-31-zephyr-3-5.md:
```diff
@@ -0,0 +1,134 @@
+
+## Upcoming Changes
+
+Moving to Zephyr 3.5 will unblock several exciting efforts that were dependant on that Zephyr release.
```
```suggestion
Moving to Zephyr 3.5 will unblock several exciting efforts that were dependent on that Zephyr release.
```

caksoylar commented on docs/blog/2024-01-31-zephyr-3-5.md:
```diff
@@ -0,0 +1,134 @@
+### West Flash Support
+
+If you have a custom board for a target that has a UF2 supporting bootloader, you can easily add support for
+flashing via `west flash`. To do so, add a line to your board's `board.cmake` file like so:
```
```suggestion
flashing via `west flash`. (Note that using `west flash` isn't mandatory, you can continue to flash your controllers via the mass storage method.)
To do so, add a line to your board's `board.cmake` file like so:
```
Perhaps worth remarking, just in case someone reading this gets confused?

petejohanson replied:
I added a note, but worded it a bit more verbosely to make clear what `west flash` does under the hood. Thoughts?

caksoylar replied:
That looks better than mine 👍

petejohanson commented:
> Will `behavior_mouse_key_press.c` also need to be switched to `POST_KERNEL`?
> 
> https://github.com/zmkfirmware/zmk/blob/40c9d9fc5993f189e55403a355aec96a5c8d8aae/app/src/behaviors/behavior_mouse_key_press.c#L42
> 
> The DTS binding for the `MAX17048` sensor (...fuel gauge?) still [collides with Zephyr's](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516169561/job/20460957357#step:11:33):
> 
> ```
> devicetree error: both /home/less/osrc/zmk-src/zmk/app/module/dts/bindings/sensor/maxim,max17048.yml and /home/less/osrc/zmk-src/zmk/zephyr/dts/bindings/fuel-gauge/maxim,max17048.yaml have 'compatible: maxim,max17048' and 'on-bus: i2c'
> ```
> 
> If I [delete ZMK's DTS binding](https://github.com/lesshonor/zmk/commit/3d0d2a6de5568e918e3ad324981c92d8ab7c717f), I get a number of different results depending upon where and in what way this thing is enabled; no idea which of these outcomes is correct, if any.
> 
>     1. [Using `MAX17048` in Kconfig.defconfig](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516258743/job/20461198917) results in a successful compilation that _appears_ to use ZMK's driver exclusively.
> 
>     2. [Using `ZMK_MAX17048` in Kconfig.defconfig](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516314666/job/20461242609) (or `CONFIG_ZMK_MAX17048=y` in `uwu30_defconfig`) fails at the linker; it seemingly can't find the `zmk,battery`-compatible node?
> 
>     3. [Using `CONFIG_FUEL_GAUGE=y` and `CONFIG_MAX17048=y` in `uwu30_defconfig`](https://github.com/lesshonor/uwu30-zmk-config/actions/runs/7516310376/job/20461236910) instead also fails at the linker, but for a different reason: it pulls in both the ZMK and Zephyr drivers??
> 
> 
> ...I'm sure some of this is my not understanding Zephyr. I have no way to test any changes on hardware yet either, so I can't determine if scenario 1 works as intended by the board's creator.
> 
> (paging @zhiayang who may be able to test on actual hardware, given #1946)

Sorry, I should have read this more careful the first time. The DTS compatible collision for sure will cause us problems. I think, given the few users of the ZMK version of the sensor driver, it would make sense to change *our* `compatible` value to avoid the collision, update consumers of the driver as needed.

The Zephyr version for sure will not "just work" since it uses the new fuel gauge API that we don't at all consume (yet).

petejohanson commented:
@lesshonor Ok, I namespaced the compatible value to be `zmk,maxim-max17048` to avoid the collision, and fixed up the Kconfig a tad too. I added a section to the prelim blog post at https://deploy-preview-1995--zmk.netlify.app/blog/2024/01/31/zephyr-3-5 that notes how to change things to match. Can you give that a try?

petejohanson commented on docs/blog/2024-01-31-zephyr-3-5.md:
```diff
@@ -0,0 +1,143 @@
+does not yet consume. To avoid a conflict with the new upstream and keep our existing sensor driver, our driver has been renamed to be namespaced with a ZMK prefix. The following changes are needed for any boards using the driver:
+
+- Change the `compatible` value for the node to be `zmk,maxim-17048`, e.g. `compatible = "zmk,maxim-max17048";`.
+- If enabling the driver explicitly via Kconfig, rename `MAXIM_MAX17048` to the new `ZMK_MAXIM_MAX17048` in your `Kconfig.defconfig` or `<board>_defconfig` files.
```
Indeed. Fixed.

petejohanson commented:
> > @lesshonor Ok, I namespaced the compatible value to be `zmk,maxim-max17048` to avoid the collision, and fixed up the Kconfig a tad too. I added a section to the prelim blog post at https://deploy-preview-1995--zmk.netlify.app/blog/2024/01/31/zephyr-3-5 that notes how to change things to match. Can you give that a try?
> 
> Thank you. I see no problem with changing the `compatible` value to `"zmk,maxim-max17048"`
> 
> https://github.com/zmkfirmware/zmk/blob/68162bc270a4d3f8f5adf4a0ecbdebb42bd9c5c6/app/module/drivers/sensor/CMakeLists.txt#L6
> 
> also needs to be changed to `CONFIG_ZMK_MAX17048`, if I'm understanding the intent of the Kconfig changes correctly.
> 
> After updating the `compatible` value and making the above change to `app/module/drivers/sensor/CMakeLists.txt`, I was able to compile both the out-of-tree board and a rebased #1946.

Indeed. Missed that spot. Thanks. Just pushed the fix.

dhruvinsh commented:
Manage to build both of my keyboard firmware against 3.5 easily. They are located [here](https://github.com/dhruvinsh/zmk-config)

This is some new warning message I observed,
```bash
warning: Deprecated symbol NRF_STORE_REBOOT_TYPE_GPREGRET is enabled.
```

Existing warning mesasges,
```
warning: ZMK_USB (defined at /workspaces/zmk/app/Kconfig:118) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: (!ZMK_SPLIT || (ZMK_SPLIT &&
ZMK_SPLIT_ROLE_CENTRAL)) (=n). See http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_ZMK_USB
and/or look up ZMK_USB in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.
```

arturo182 commented on app/boards/arm/puchi_ble/pinmux.c:
```diff
@@ -11,9 +11,7 @@
-static int pinmux_puchi_ble_init(const struct device *port) {
-    ARG_UNUSED(port);
-
+static int pinmux_puchi_ble_init() {
```
I noticed that in other places you put `void` for empty parameters list (as is tradition), but not here. You might want to add it for consistency. Maybe that should be a pre-commit hook as well? Unsure, haven't used pre-commit before.

petejohanson replied:
Indeed. Good catch. Fixed.

petejohanson commented:
Feeling generally positive for merging this. One oustanding report if pairing issues on Arch Linux, waiting for the tester to report ZMK logs to see what's up. They are using the MoErgo version, so need to be careful this isn't related to any of their customizations.

caksoylar commented on docs/docs/development/new-behavior.mdx:
```diff
@@ -116,16 +116,16 @@ These are additional variables required to configure a particular instance of a
 
 :::info
-For more information on additional `properties`, refer to [Zephyr's documentation on Devicetree bindings](https://docs.zephyrproject.org/latest/build/dts/bindings.html#properties).
+For more information on additional `properties`, refer to [Zephyr's documentation on Devicetree bindings](https://docs.zephyrproject.org/3.5.0/build/dts/bindings.html#properties).
```
```suggestion
For more information on additional `properties`, refer to [Zephyr's documentation on Devicetree bindings](https://docs.zephyrproject.org/3.5.0/build/dts/bindings-syntax.html#properties).
```

caksoylar commented on docs/docs/development/new-shield.mdx:
```diff
@@ -25,7 +25,7 @@ The high level steps are:
 - Add support for features such as encoders, OLED displays, or RGB underglow.
 
-It may be helpful to review the upstream [shields documentation](https://docs.zephyrproject.org/3.2.0/guides/porting/shields.html#shields) to get a proper understanding of the underlying system before continuing.
+It may be helpful to review the upstream [shields documentation](https://docs.zephyrproject.org/3.5.0/guides/porting/shields.html#shields) to get a proper understanding of the underlying system before continuing.
```
```suggestion
It may be helpful to review the upstream [shields documentation](https://docs.zephyrproject.org/3.5.0/hardware/porting/shields.html#shields) to get a proper understanding of the underlying system before continuing.
```

caksoylar commented on docs/docs/faq.md:
```diff
@@ -7,10 +7,10 @@ sidebar_label: FAQs
 - Optimization for low-powered, small memory footprint devices.
-- Powerful hardware abstraction and configuration using [DeviceTree](https://docs.zephyrproject.org/latest/guides/dts/index.html) and [Kconfig](https://docs.zephyrproject.org/latest/build/kconfig/index.html).
-- A BLE stack that periodically obtains [qualification](https://docs.zephyrproject.org/latest/connectivity/bluetooth/bluetooth-qual.html) listings, making it easier for final products to obtain qualification from the Bluetooth® SIG.
+- Powerful hardware abstraction and configuration using [DeviceTree](https://docs.zephyrproject.org/3.5.0/guides/dts/index.html) and [Kconfig](https://docs.zephyrproject.org/3.5.0/build/kconfig/index.html).
```
```suggestion
- Powerful hardware abstraction and configuration using [DeviceTree](https://docs.zephyrproject.org/3.5.0/build/dts/index.html) and [Kconfig](https://docs.zephyrproject.org/3.5.0/build/kconfig/index.html).
```

caksoylar reviewed (changes requested):
Did a pass over the updated docs links, some of these are auto-redirecting but probably good to update nevertheless.

arturo182 commented:
![giphy (2)](https://github.com/zmkfirmware/zmk/assets/249082/1cd8cd5e-84bf-4a86-a5c5-481c6b2d9dce)
